connector/src/main/scala/com/microsoft/kusto/spark/datasource/KustoRelation.scala

```diff
@@ -38,79 +38,77 @@ private[kusto] case class KustoRelation(kustoCoordinates: KustoCoordinates,
-        case e: Exception =>
-          if (forceSingleMode) {
-            // Throw in case user forced LeanMode
-            throw e
-          }
-          // By default we fallback to distributed mode
-          timedOutCounting = true
-      }
-      if (count == 0 && !timedOutCounting) {
-        res = Some(sparkSession.emptyDataFrame.rdd)
-      } else {
-        // Use distributed mode if count is high or in case of a time out
-        useSingleMode =  !(timedOutCounting || count > KustoConstants.DirectQueryUpperBoundRows)
-      }
+    val kustoClient = KustoClientCache.getClient(kustoCoordinates.clusterUrl, authentication,
+      kustoCoordinates.ingestionUrl, kustoCoordinates.clusterAlias)
+    val isUserOptionForceSingleMode = readOptions.readMode.contains(ReadMode.ForceSingleMode)
+    val storageParameters = maybeStorageParameters.getOrElse(kustoClient.getTempBlobsForExport)
+    val (useSingleMode, estimatedRecordCount) = readOptions.readMode match {
+      // if the user provides a specific option , this is to be used no matter what
+      case Some(_) => (isUserOptionForceSingleMode, -1)
+      // If there is no option mentioned , then estimate which option to use
+      // Count records and see if we wat a distributed or single mode
+      case None => val estimatedRecordCountResult =  Try(KDSU.
+        estimateRowsCount(kustoClient.engineClient, query, kustoCoordinates.database, clientRequestProperties.orNull))
+        estimatedRecordCountResult match {
+          // if the count is lss than the 5k threshold,use single mode.
+          case Success(recordCount) => (recordCount <= KustoConstants.DirectQueryUpperBoundRows, recordCount)
+          // A case of query timing out. ForceDistributedMode will be used here
+          case Failure(_) => (false, -1)
+        }
     }
-
-    var exception: Option[Exception] = None
-    if(res.isEmpty) {
-      if (useSingleMode) {
-        try {
-          res = Some(KustoReader.singleBuildScan(
-            kustoClient,
-            KustoReadRequest(sparkSession, cachedSchema, kustoCoordinates, query, authentication, timeout, clientRequestProperties, requestId),
-            KustoFiltering(requiredColumns, filters)))
-        } catch {
-          case ex: Exception => exception = Some(ex)
+    // if the selection yields 0 rows there are no records to scan and process
+    if(estimatedRecordCount == 0){
+      sparkSession.emptyDataFrame.rdd
+    } else {
+      // either a case of non-zero records or a case of timed-out.`
+      if(useSingleMode){
+        // there are less than 5000 (KustoConstants.DirectQueryUpperBoundRows) records, perform a single scan
+        Try(KustoReader.singleBuildScan(
+          kustoClient,
+          KustoReadRequest(sparkSession, cachedSchema, kustoCoordinates, query, authentication, timeout, clientRequestProperties, requestId),
+          KustoFiltering(requiredColumns, filters))) match {
+          case Success(rdd) => rdd
```

Comment:
I think your indentation with 'Try match'  is confusing
please align the word 'case' with Try 
or new line the match and align it with Try

Reply:
Split this into readable lines